crates/astria-sequencer/src/transaction/unsigned.rs

```diff
@@ -54,21 +79,86 @@ impl Unsigned {
     }
 }
 
+#[async_trait::async_trait]
+impl ActionHandler for Unsigned {
+    fn check_stateless(&self) -> Result<()> {
+        for action in &self.actions {
+            match action {
+                Action::TransferAction(tx) => tx.check_stateless()?,
+                Action::SequenceAction(tx) => tx.check_stateless()?,
+            }
+        }
+        Ok(())
+    }
+
+    async fn check_stateful<S: StateReadExt + 'static>(
+        &self,
+        state: &S,
+        from: &Address,
+    ) -> Result<()> {
+        let curr_nonce = state.get_account_nonce(from).await?;
+        ensure!(curr_nonce < self.nonce, "invalid nonce");
+
+        // do we need to make a StateDelta here so we can check the actions on the successive state?
+        for action in &self.actions {
+            match action {
+                Action::TransferAction(tx) => tx.check_stateful(state, from).await?,
+                Action::SequenceAction(tx) => tx.check_stateful(state, from).await?,
+            }
+        }
+
+        Ok(())
+    }
+
+    #[instrument(
+        skip_all,
+        fields(
+            nonce = self.nonce.into_inner(),
```

Comment:
Also print the address as a base64 field if applicable? I think it already implements Display?